### PR TITLE
Example to use classnames

### PR DESCRIPTION
**Pull request by dai-shi**
Fix `module is not defined` error when we used `classnames` on the server.
The fix doesn't seem ideal. I would like to hear other options from vite experts.

### CODE REVIEW

**Review by himself65 — COMMENTED** [review]
The reason is Vite dev doesn't support the CJS module, so you have to set all packages to ESM.

So it's not a good idea to add more externals.

**Comment by dai-shi** [comment]
@himself65 What would be other workarounds for `classnames`?

> What if you use clsx?

It works. #359

**Comment by himself65** [comment]
https://www.npmjs.com/package/vite-plugin-commonjs would be a good one?

**Comment by dai-shi** [comment]
Thanks. Let me try.

**Comment by dai-shi** [comment]
> https://www.npmjs.com/package/vite-plugin-commonjs would be a good one?

Unfortunately, it doesn't work out of the box. Would you like to give a try? @himself65

**Comment by himself65** [comment]
Let me check

**Review by himself65 — CHANGES_REQUESTED** [review]
This works for me

```diff
Index: examples/12_css/package.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/examples/12_css/package.json b/examples/12_css/package.json
--- a/examples/12_css/package.json	(revision aadbe84b2ba4787407d1591a195a0b785eacdbfc)
+++ b/examples/12_css/package.json	(date 1704910250658)
@@ -22,6 +22,7 @@
     "@vanilla-extract/vite-plugin": "3.9.3",
     "typescript": "5.3.3",
     "vite": "5.0.10",
+    "vite-plugin-commonjs": "^0.10.1",
     "vite-plugin-stylex": "0.4.1"
   }
 }
Index: examples/12_css/vite.config.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/examples/12_css/vite.config.ts b/examples/12_css/vite.config.ts
--- a/examples/12_css/vite.config.ts	(revision aadbe84b2ba4787407d1591a195a0b785eacdbfc)
+++ b/examples/12_css/vite.config.ts	(date 1704910429912)
@@ -1,10 +1,23 @@
 import { defineConfig } from 'vite';
 import { vanillaExtractPlugin } from '@vanilla-extract/vite-plugin';
 import styleX from 'vite-plugin-stylex';
+import commonjs from 'vite-plugin-commonjs';
 
 export default defineConfig({
   ssr: {
     external: ['@stylexjs/stylex'],
   },
-  plugins: [vanillaExtractPlugin({ emitCssInSsr: true }), styleX()],
+  plugins: [
+    vanillaExtractPlugin({ emitCssInSsr: true }),
+    styleX(),
+    commonjs({
+      filter(id: string) {
+        // `node_modules` is exclude by default, so we need to include it explicitly
+        // https://github.com/vite-plugin/vite-plugin-commonjs/blob/v0.7.0/src/index.ts#L125-L127
+        if (id.includes('node_modules/classnames')) {
+          return true;
+        }
+      },
+    }),
+  ],
 });
Index: packages/waku/src/lib/handlers/dev-worker-impl.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/waku/src/lib/handlers/dev-worker-impl.ts b/packages/waku/src/lib/handlers/dev-worker-impl.ts
--- a/packages/waku/src/lib/handlers/dev-worker-impl.ts	(revision aadbe84b2ba4787407d1591a195a0b785eacdbfc)
+++ b/packages/waku/src/lib/handlers/dev-worker-impl.ts	(date 1704910371209)
@@ -130,8 +130,7 @@
       conditions: ['react-server', 'workerd'],
       externalConditions: ['react-server', 'workerd'],
     },
-    // FIXME adding `classnames` here seems very ad-hoc.
-    external: ['react', 'react-server-dom-webpack', 'classnames'],
+    external: ['react', 'react-server-dom-webpack'],
     noExternal: /^(?!node:)/,
   },
   appType: 'custom',
Index: examples/12_css/tsconfig.node.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/examples/12_css/tsconfig.node.json b/examples/12_css/tsconfig.node.json
--- a/examples/12_css/tsconfig.node.json	(revision aadbe84b2ba4787407d1591a195a0b785eacdbfc)
+++ b/examples/12_css/tsconfig.node.json	(date 1704910331508)
@@ -2,7 +2,8 @@
   "extends": "./tsconfig.json",
   "compilerOptions": {
     "outDir": "./dist/node",
-    "types": ["node"]
+    "types": ["node"],
+    "moduleResolution": "bundler"
   },
   "include": ["vite.config.ts"]
 }
Index: pnpm-lock.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pnpm-lock.yaml b/pnpm-lock.yaml
--- a/pnpm-lock.yaml	(revision aadbe84b2ba4787407d1591a195a0b785eacdbfc)
+++ b/pnpm-lock.yaml	(date 1704910250580)
@@ -434,6 +434,9 @@
       vite:
         specifier: 5.0.10
         version: 5.0.10(@types/node@20.10.6)
+      vite-plugin-commonjs:
+        specifier: ^0.10.1
+        version: 0.10.1
       vite-plugin-stylex:
         specifier: 0.4.1
         version: 0.4.1(@babel/traverse@7.23.7)(@babel/types@7.23.6)(vite@5.0.10)
@@ -4349,6 +4352,13 @@
     dependencies:
       yallist: 4.0.0
     dev: true
+
+  /magic-string@0.30.5:
+    resolution: {integrity: sha512-7xlpfBaQaP/T6Vh8MO/EqXSW5En6INHEvEXQiuff7Gku0PWjU3uf6w/j9o7O+SpB5fOAkrI5HeoNgwjEO0pFsA==}
+    engines: {node: '>=12'}
+    dependencies:
+      '@jridgewell/sourcemap-codec': 1.4.15
+    dev: true
 
   /markdown-extensions@1.1.1:
     resolution: {integrity: sha512-WWC0ZuMzCyDHYCasEGs4IPvLyTGftYwh6wIEOULOF0HXcqZlhwRzrK0w2VUlxWA98xnvb/jszw4ZSkJ6ADpM6Q==}
@@ -6408,6 +6418,24 @@
       - supports-color
       - terser
     dev: true
+
+  /vite-plugin-commonjs@0.10.1:
+    resolution: {integrity: sha512-taP8R9kYGlCW5OzkVR0UIWRCnG6rSxeWWuA7tnU5b9t5MniibOnDY219NhisTeDhJAeGT8cEnrhVWZ9A5yD+vg==}
+    dependencies:
+      acorn: 8.11.3
+      fast-glob: 3.3.2
+      magic-string: 0.30.5
+      vite-plugin-dynamic-import: 1.5.0
+    dev: true
+
+  /vite-plugin-dynamic-import@1.5.0:
+    resolution: {integrity: sha512-Qp85c+AVJmLa8MLni74U4BDiWpUeFNx7NJqbGZyR2XJOU7mgW0cb7nwlAMucFyM4arEd92Nfxp4j44xPi6Fu7g==}
+    dependencies:
+      acorn: 8.11.3
+      es-module-lexer: 1.4.1
+      fast-glob: 3.3.2
+      magic-string: 0.30.5
+    dev: true
 
   /vite-plugin-stylex@0.4.1(@babel/traverse@7.23.7)(@babel/types@7.23.6)(vite@5.0.10):
     resolution: {integrity: sha512-QBkZYZyf1Hj1pUGYejF2rR3nuZyHt3YrkwPPu3dr/algyurtps+YTJ5Db2zVSd8sSPkHURT9UB1CQaJYnMshWg==}
```

File: examples/12_css/vite.config.ts [hunk 1]
```diff
@@ -1,10 +1,24 @@
...
   },
-  plugins: [vanillaExtractPlugin({ emitCssInSsr: true }), styleX()],
+  plugins: [
+    vanillaExtractPlugin({ emitCssInSsr: true }),
+    styleX(),
+    // @ts-expect-error not callable FIXME why not callable?
+    commonjs({
+      filter(id: string) {
+        // `node_modules` is exclude by default, so we need to include it explicitly
+        // https://github.com/vite-plugin/vite-plugin-commonjs/blob/v0.7.0/src/index.ts#L125-L127
```
Comment (dai-shi) [inline examples/12_css/vite.config.ts:17]:
```suggestion
        // https://github.com/vite-plugin/vite-plugin-commonjs/blob/v0.10.1/src/index.ts#L141-L142
```

**Comment by dai-shi** [comment]
Let's revisit it and see if there's a better solution in the framework side in the future. For now, this is a workaround to use classnames.

**Comment by dai-shi** [comment]
> The reason is Vite dev doesn't support the CJS module, so you have to set all packages to ESM.
> 
> So it's not a good idea to add more externals.

@himself65 I still think externalizing is a valid "workaround": https://github.com/dai-shi/waku/issues/110#issuecomment-1891208899

Do you only mean it's a bad idea to list random libraries in the waku core source code?

**Comment by himself65** [comment]
> Do you only mean it's a bad idea to list random libraries in the waku core source code?

I'm not sure what happened in prod if you set ssr external. I guess it will require `node_modules` in the runtime.

**Comment by dai-shi** [comment]
prd works without the workaround (`exteranl: ...`). I don't know why, actually.

See this is dev-only config:
https://github.com/dai-shi/waku/blob/7398b45ed1c7858e3a24c04c35fd32efc2bc54b2/packages/website/vite.config.ts#L4-L10

**Pull request merged**